### PR TITLE
[CoreBundle ] Api v2

### PR DESCRIPTION
**Pull request by ngodfraind**
| Q             | A
| ------------- | ---
| Bug fix?      | yes/no
| New feature?  | yes/no
| BC breaks?    | yes/no
| Fixed tickets | comma-separated list of tickets fixed by the PR, if any

TODO

- [x] use doctrine annotation for default serialization
- [x] remove traits from AbstractController
- [x] better handling of serializer options @Elorfin 

NICE TO HAVE BECAUSE IT'S DIFFICULT

- [ ] add Uuid for workspace
- [ ] migrate Guid for workspace
- [ ] rename mail into email

CAREFUL WITH USER UUID

chat might be broken
cursus might be broken

### CODE REVIEW

File: main/core/API/Crud/CrudRole.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+use Claroline\CoreBundle\Entity\Role;
+use Claroline\CoreBundle\Event\CrudEvent;
+use JMS\DiExtraBundle\Annotation as DI;
+
+/**
+ * @DI\Service("claroline.crud.role")
+ * @DI\Tag("claroline.crud")
+ */
+class CrudRole
```
Comment (Elorfin) [inline main/core/API/Crud/CrudRole.php]: it should be `RoleCrud`.

File: main/core/API/Finder/GroupFinder.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            $qb->leftJoin('obj.organizations', 'uo');
+            $qb->leftJoin('uo.administrators', 'ua');
+            $qb->andWhere('ua.id = :userId');
+            $qb->setParameter('userId', $currentUser->getId());
+        }
+
+        foreach ($searches as $filterName => $filterValue) {
+            switch ($filterName) {
+              default:
+                if ('true' === $filterValue || 'false' === $filterValue || true === $filterValue || false === $filterValue) {
```
Comment (Elorfin) [inline main/core/API/Finder/GroupFinder.php]: for @Elorfin : transtyping is done earlier in new Finder version.

File: main/core/API/Finder/RoleFinder.php [hunk 3]
```diff
@@ -31,19 +58,43 @@ public function configureQueryBuilder(QueryBuilder $qb, array $searches = [])
...
+
               default:
                 if ('true' === $filterValue || 'false' === $filterValue || true === $filterValue || false === $filterValue) {
                     $filterValue = is_string($filterValue) ? 'true' === $filterValue : $filterValue;
                     $qb->andWhere("obj.{$filterName} = :{$filterName}");
                     $qb->setParameter($filterName, $filterValue);
                 } else {
-                    $qb->andWhere("UPPER(obj.{$filterName}) LIKE :{$filterName}");
-                    $qb->setParameter($filterName, '%'.strtoupper($filterValue).'%');
+                    if (is_int($filterValue)) {
```
Comment (Elorfin) [inline main/core/API/Finder/RoleFinder.php]: for @Elorfin : transtyping is done earlier in new Finder version.

File: main/core/API/Finder/UserFinder.php [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            $qb->leftJoin('obj.organizations', 'uo');
+            $qb->leftJoin('uo.administrators', 'ua');
+            $qb->andWhere('ua.id = :userId');
+            $qb->setParameter('userId', $currentUser->getId());
+        }
+
+        foreach ($searches as $filterName => $filterValue) {
+            switch ($filterName) {
+              default:
+                if ('true' === $filterValue || 'false' === $filterValue || true === $filterValue || false === $filterValue) {
```
Comment (Elorfin) [inline main/core/API/Finder/UserFinder.php]: for @Elorfin : transtyping is done earlier in new Finder version.

File: main/core/API/Serializer/AbstractSerializer.php [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+        if (isset($data->id) || isset($data->uuid)) {
+            if (isset($data->uuid)) {
+                $object = $this->om->getRepository($class)->findOneByUuid($data->uuid);
+            } else {
+                $object = is_int($data->id) ?
+                $this->om->getRepository($class)->findOneById($data->id) :
+                $this->om->getRepository($class)->findOneByUuid($data->id);
+            }
+        }
```
Comment (Elorfin) [inline main/core/API/Serializer/AbstractSerializer.php]: +1

File: main/core/API/SerializerProvider.php [hunk 6]
```diff
@@ -78,4 +78,16 @@ public function serialize($object, $options = [])
...
+    /**
+     * Serializes an object.
+     *
+     * @param $object - the object to serialize
+     *
+     * @return mixed - a json serializable structure
+     */
+    public function deserialize($class, $data, $options = [])
+    {
+        return $this->get($class)->deserialize($class, $data);
```
Comment (Elorfin) [inline main/core/API/SerializerProvider.php]: `options` array is not passed to the serializer instance.

File: main/core/Routing/ApiLoader.php [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+        $file = $this->locator->locate($resource);
+
+        $routes = new RouteCollection();
+
+        $defaults = [
+          'create' => ['', 'POST'],
+          'update' => ['{uuid}', 'PUT'],
+          'deleteBulk' => ['', 'DELETE'],
+          'list' => ['{page}/{limit}', 'GET'],
```
Comment (Elorfin) [inline main/core/Routing/ApiLoader.php]: for @Elorfin : new Finder version handles page and limit from queryString

File: main/core/API/Finder/RoleFinder.php (hunk 3 above)
Comment (ngodfraind) [reply main/core/API/Finder/RoleFinder.php]: that one isn't transtyping. We're just checking if it's an int or a string in order to permform a LIKE or sthg else in the request.

**Pull request merged**